src/librustc_mir/transform/instcombine.rs

```diff
@@ -43,12 +43,20 @@ impl<'tcx> MutVisitor<'tcx> for InstCombineVisitor<'tcx> {
             let new_place = match *rvalue {
                 Rvalue::Ref(_, _, Place {
                     ref mut base,
-                    projection: Some(ref mut projection),
-                }) => Place {
-                    // Replace with dummy
-                    base: mem::replace(base, PlaceBase::Local(Local::new(0))),
-                    projection: projection.base.take(),
-                },
+                    projection: ref mut projection @ box [.., _],
+                }) => {
+                    let mut projection1 = projection.to_vec();
+                    let projection2 = projection1.split_off(projection1.len() - 1);
```

Comment:
split the slice and do the `to_vec` at the use sites. This way we don't create an alloc of n elements, cut that to n-1 elements and then have `into_boxed_slice` create an alloc of n-1 elements and trash the old one